### PR TITLE
Fix overflow check

### LINKED ISSUE #276: [boolector] boolector_add: sorts of 'e0' and 'e1' must match
ESBMC crashes when verifying this C program:

```
long nondet_long(void);
 
int main(void) {
  long l = 0;
  int c = 0;
  long t = nondet_long();
  long s = nondet_long();
  long *b = &s;
 
  l = (t + *b) & (0xffffffffL); 
  c += (l < t);
}
```

```
$ esbmc ari.c --overflow-check
ESBMC version 6.4.0 64-bit x86_64 macos
file ari.c: Parsing
Converting
Generating GOTO Program
GOTO program creation time: 5.380s
GOTO program processing time: 0.050s
Starting Bounded Model Checking
Symex completed in: 0.109s (21 assignments)
Slicing time: 0.000s (removed 14 assignments)
Generated 6 VCC(s), 2 remaining after simplification (7 assignments)
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Boolector error encountered
[boolector] boolector_add: sorts of 'e0' and 'e1' must match


Abort trap: 6
```

### PR DESCRIPTION
**Pull request by lucasccordeiro**
This PR aims to fix ill-formed expressions to check for arithmetic overflow.

### CODE REVIEW

File: src/goto-programs/goto_check.cpp [hunk 1]
```diff
@@ -212,10 +212,32 @@ void goto_checkt::overflow_check(
...
+  {
+    typecast2tc casted_size_2(side_1->type, side_2);
+    if(is_add2t(expr))
+      new_expr = add2tc(side_1->type, side_1, casted_size_2);
+    else if(is_sub2t(expr))
+      new_expr = sub2tc(side_1->type, side_1, casted_size_2);
+    else if(is_mul2t(expr))
+      new_expr = mul2tc(side_1->type, side_1, casted_size_2);
+    else if(is_shl2t(expr))
+      new_expr = shl2tc(side_1->type, side_1, casted_size_2);
```
Comment (rafaelsamenezes) [inline src/goto-programs/goto_check.cpp]: Instead of `if...else-if...else` I think that using a switch-case with `expr2t::expr_ids` would be better.

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

**Review by rafaelsamenezes — APPROVED** [review]
LGTM

**Review by mikhailramalho — COMMENTED** [review]
The problem here is that by always casting the `side2` to the type of the `side1`, we might violate type promotion rules and miss bugs, i.e., if `side2` is greater than `side1`, it is the `side1` that needs to be promoted to the type of `side2`.

What we are missing here is the type promotion by the frontend, i.e, esbmc prints:
```
        c=c + (l < t);
```
while cbmc prints:
```
        c = c + (signed int)(l < t);
```

If I had to guess, the issue it because I never ported `c_typecheck_baset::typecheck_expr_pointer_arithmetic` from the cbmc's backend (removed in d2d47fa291). Here's the cbmc's implementation: `https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/c_typecheck_expr.cpp#L3271`

**Comment by mikhailramalho** [comment]
Could you open a new PR (or remove the unnecessary commits)? So we don't push unnecessary code to the repo.

**Comment by lucasccordeiro** [comment]
Closed via https://github.com/esbmc/esbmc/pull/279.

**Pull request closed without merging**